app/src/ingestion/markdown_chunking.py

```diff
@@ -0,0 +1,341 @@
+            )
+            # Why check for are_tokens_frozen? Because calling copy() on Paragraph tokens doesn't work.
+            # Fortunately if we use "freeze_token_children", then we don't need to copy Paragraph tokens
+            if not are_tokens_frozen:
+                n.data.token = copy(n.data.token)
+
+    assert subtree[node.data_id], f"Expected data_id {node.data_id!r} for {subtree.first_child()}"
+
+    # Now that copies of node.data and node.data.token are created, update references to the tokens
+    # Update all node.data.token.children to point to the new token objects in the subtree
+    for n in new_node.iterator(add_self=True):
+        if isinstance(n.data, TokenNodeData) and not n.data["freeze_token_children"]:
+            n.data.token.children = [
+                c.token for c in n.children if isinstance(c.data, TokenNodeData)
+            ]
+    # At this point, no object in the subtree should be pointing to objects in the original tree,
+    # except for tokens associated with "freeze_token_children". We are free to modify the subtree.
+    return subtree
+
+
+def remove_child(node: Node, child: Node) -> None:
+    logger.info("Removing child %s from %s", child.data_id, node.data_id)
+    # Update node.token.children since that's used for rendering
+    node.data.token.children.remove(child.data.token)
+    # Then remove the child from the tree
+    child.remove()
+
+
+def remove_children_from(node: Node, child_data_ids: set[str]) -> None:
+    "Remove children nodes with data_id in child_data_ids from node"
+    # Do not remove children while iterating through them
+    # Create a list of child nodes to remove, then remove them
+    nodes_to_remove = [c for c in node.children if c.data_id in child_data_ids]
+    if len(nodes_to_remove) != len(child_data_ids):
+        logger.warning(
+            "Expected to remove %s, but found only %s",
+            child_data_ids,
+            set(n.data_id for n in nodes_to_remove),
+        )
+    for child_node in nodes_to_remove:
+        remove_child(node, child_node)
+
+
+def find_closest_ancestor(
+    node: Node, does_match: Callable[[Node], bool], include_self: bool = False
+) -> Node:
+    "Return the first parent/ancestor node that matches the does_match function"
+    if include_self and does_match(node):
+        return node
+
+    while node.parent:
+        if does_match(node.parent):
+            return node.parent
+        node = node.parent
+    return None
+
+
+def nodes_as_markdown(nodes: Sequence[Node]) -> str:
+    # Don't render Heading nodes by themselves
+    if len(nodes) == 1 and nodes[0].data_type in ["Heading"]:
+        return ""
+
+    md_list: list[str] = []
+    for node in nodes:
+        # node.data["summary"] is set when node is chunked
+        node_md = node.data["summary"] or render_subtree_as_md(node, normalize=True)
+        md_list.append(node_md)
+    return normalize_markdown("".join(md_list))
+
+
+@dataclass
+class ProtoChunk:
```

Comment:
Not sure what I'll do with this class. It's useful for unit tests and could be moved there. For the most part, only `ChunkingConfig` uses it (by design).
My next PR will be integrating this into EDD website ingestion, so I'll decide at that point.